{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "01dd3913",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(175321, 14)"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "dataset_df = pd.read_csv(\"data/TableS1_HQ_filtered16S.csv\")\n",
    "dataset_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "17c5f49a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly_accession</th>\n",
       "      <th>CheckM completeness</th>\n",
       "      <th>CheckM contamination</th>\n",
       "      <th>16S_ID</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>Sum</th>\n",
       "      <th>PKSI</th>\n",
       "      <th>PKSother</th>\n",
       "      <th>NRPS</th>\n",
       "      <th>RiPPs</th>\n",
       "      <th>Saccharides</th>\n",
       "      <th>Terpene</th>\n",
       "      <th>PKS-NRP_Hybrids</th>\n",
       "      <th>Others</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_000022305.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.67</td>\n",
       "      <td>refseq4905</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_000025965.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.37</td>\n",
       "      <td>refseq15376</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_000023105.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.84</td>\n",
       "      <td>refseq45559</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_000012365.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.06</td>\n",
       "      <td>refseq128457</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_000024165.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.88</td>\n",
       "      <td>refseq35003</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  assembly_accession  CheckM completeness  CheckM contamination        16S_ID  \\\n",
       "0    GCF_000022305.1                100.0                  0.67    refseq4905   \n",
       "1    GCF_000025965.1                100.0                  0.37   refseq15376   \n",
       "2    GCF_000023105.1                100.0                  0.84   refseq45559   \n",
       "3    GCF_000012365.1                100.0                  0.06  refseq128457   \n",
       "4    GCF_000024165.1                100.0                  0.88   refseq35003   \n",
       "\n",
       "    assembly_level  Sum  PKSI  PKSother  NRPS  RiPPs  Saccharides  Terpene  \\\n",
       "0  Complete Genome    3     0         0     0      0            0        1   \n",
       "1  Complete Genome    7     0         0     0      0            0        1   \n",
       "2  Complete Genome    1     0         0     0      0            0        1   \n",
       "3  Complete Genome   20     1         0     5      3            0        2   \n",
       "4  Complete Genome    8     0         0     2      1            0        1   \n",
       "\n",
       "   PKS-NRP_Hybrids  Others  \n",
       "0                0       2  \n",
       "1                0       6  \n",
       "2                0       0  \n",
       "3                2       7  \n",
       "4                0       4  "
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "e4b429bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly_accession</th>\n",
       "      <th>rna_accession</th>\n",
       "      <th>id</th>\n",
       "      <th>rna_length</th>\n",
       "      <th>sum_bgc</th>\n",
       "      <th>taxid</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>kindom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_000003135.1</td>\n",
       "      <td>NZ_GG666857.1_rrna_62</td>\n",
       "      <td>1289</td>\n",
       "      <td>1535</td>\n",
       "      <td>0</td>\n",
       "      <td>548480</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Actinobacteria</td>\n",
       "      <td>c__Actinomycetia</td>\n",
       "      <td>o__Bifidobacteriales</td>\n",
       "      <td>f__Bifidobacteriaceae</td>\n",
       "      <td>g__Bifidobacterium</td>\n",
       "      <td>s__Bifidobacterium longum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_000003215.1</td>\n",
       "      <td>NZ_ABFD02000028.1_rrna_53</td>\n",
       "      <td>66</td>\n",
       "      <td>1503</td>\n",
       "      <td>2</td>\n",
       "      <td>455631</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Firmicutes</td>\n",
       "      <td>c__Clostridia</td>\n",
       "      <td>o__Eubacteriales</td>\n",
       "      <td>f__Peptostreptococcaceae</td>\n",
       "      <td>g__Clostridioides</td>\n",
       "      <td>s__Clostridioides difficile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_000003645.1</td>\n",
       "      <td>KEECGHFP_00466</td>\n",
       "      <td>40703</td>\n",
       "      <td>849</td>\n",
       "      <td>6</td>\n",
       "      <td>526973</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Firmicutes</td>\n",
       "      <td>c__Bacilli</td>\n",
       "      <td>o__Bacillales</td>\n",
       "      <td>f__Bacillaceae</td>\n",
       "      <td>g__Bacillus</td>\n",
       "      <td>s__Bacillus cereus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_000003925.1</td>\n",
       "      <td>NZ_CM000742.1_rrna_3</td>\n",
       "      <td>108</td>\n",
       "      <td>1552</td>\n",
       "      <td>11</td>\n",
       "      <td>526997</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Firmicutes</td>\n",
       "      <td>c__Bacilli</td>\n",
       "      <td>o__Bacillales</td>\n",
       "      <td>f__Bacillaceae</td>\n",
       "      <td>g__Bacillus</td>\n",
       "      <td>s__Bacillus mycoides</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_000003955.1</td>\n",
       "      <td>NZ_CM000741.1_rrna_101</td>\n",
       "      <td>116779</td>\n",
       "      <td>1643</td>\n",
       "      <td>8</td>\n",
       "      <td>526994</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Firmicutes</td>\n",
       "      <td>c__Bacilli</td>\n",
       "      <td>o__Bacillales</td>\n",
       "      <td>f__Bacillaceae</td>\n",
       "      <td>g__Bacillus</td>\n",
       "      <td>s__Bacillus cereus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  assembly_accession              rna_accession      id  rna_length  sum_bgc  \\\n",
       "0    GCF_000003135.1      NZ_GG666857.1_rrna_62    1289        1535        0   \n",
       "1    GCF_000003215.1  NZ_ABFD02000028.1_rrna_53      66        1503        2   \n",
       "2    GCF_000003645.1             KEECGHFP_00466   40703         849        6   \n",
       "3    GCF_000003925.1       NZ_CM000742.1_rrna_3     108        1552       11   \n",
       "4    GCF_000003955.1     NZ_CM000741.1_rrna_101  116779        1643        8   \n",
       "\n",
       "    taxid assembly_level       kindom             phylum             class  \\\n",
       "0  548480       Scaffold  k__Bacteria  p__Actinobacteria  c__Actinomycetia   \n",
       "1  455631     Chromosome  k__Bacteria      p__Firmicutes     c__Clostridia   \n",
       "2  526973     Chromosome  k__Bacteria      p__Firmicutes        c__Bacilli   \n",
       "3  526997     Chromosome  k__Bacteria      p__Firmicutes        c__Bacilli   \n",
       "4  526994     Chromosome  k__Bacteria      p__Firmicutes        c__Bacilli   \n",
       "\n",
       "                  order                    family               genus  \\\n",
       "0  o__Bifidobacteriales     f__Bifidobacteriaceae  g__Bifidobacterium   \n",
       "1      o__Eubacteriales  f__Peptostreptococcaceae   g__Clostridioides   \n",
       "2         o__Bacillales            f__Bacillaceae         g__Bacillus   \n",
       "3         o__Bacillales            f__Bacillaceae         g__Bacillus   \n",
       "4         o__Bacillales            f__Bacillaceae         g__Bacillus   \n",
       "\n",
       "                       species  \n",
       "0    s__Bifidobacterium longum  \n",
       "1  s__Clostridioides difficile  \n",
       "2           s__Bacillus cereus  \n",
       "3         s__Bacillus mycoides  \n",
       "4           s__Bacillus cereus  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lineage_df = pd.read_csv(\"data/psmpa_metadata.tsv\", sep='\\t')\n",
    "lineage_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "cfd962f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly_accession</th>\n",
       "      <th>rna_length</th>\n",
       "      <th>kindom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_000003135.1</td>\n",
       "      <td>1535</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Actinobacteria</td>\n",
       "      <td>c__Actinomycetia</td>\n",
       "      <td>o__Bifidobacteriales</td>\n",
       "      <td>f__Bifidobacteriaceae</td>\n",
       "      <td>g__Bifidobacterium</td>\n",
       "      <td>s__Bifidobacterium longum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_000003215.1</td>\n",
       "      <td>1503</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Firmicutes</td>\n",
       "      <td>c__Clostridia</td>\n",
       "      <td>o__Eubacteriales</td>\n",
       "      <td>f__Peptostreptococcaceae</td>\n",
       "      <td>g__Clostridioides</td>\n",
       "      <td>s__Clostridioides difficile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_000003645.1</td>\n",
       "      <td>849</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Firmicutes</td>\n",
       "      <td>c__Bacilli</td>\n",
       "      <td>o__Bacillales</td>\n",
       "      <td>f__Bacillaceae</td>\n",
       "      <td>g__Bacillus</td>\n",
       "      <td>s__Bacillus cereus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_000003925.1</td>\n",
       "      <td>1552</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Firmicutes</td>\n",
       "      <td>c__Bacilli</td>\n",
       "      <td>o__Bacillales</td>\n",
       "      <td>f__Bacillaceae</td>\n",
       "      <td>g__Bacillus</td>\n",
       "      <td>s__Bacillus mycoides</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_000003955.1</td>\n",
       "      <td>1643</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Firmicutes</td>\n",
       "      <td>c__Bacilli</td>\n",
       "      <td>o__Bacillales</td>\n",
       "      <td>f__Bacillaceae</td>\n",
       "      <td>g__Bacillus</td>\n",
       "      <td>s__Bacillus cereus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216403</th>\n",
       "      <td>GCF_907275085.1</td>\n",
       "      <td>88</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>o__Enterobacterales</td>\n",
       "      <td>f__Enterobacteriaceae</td>\n",
       "      <td>g__Klebsiella</td>\n",
       "      <td>s__Klebsiella pneumoniae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216404</th>\n",
       "      <td>GCF_907275095.1</td>\n",
       "      <td>299</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>o__Enterobacterales</td>\n",
       "      <td>f__Enterobacteriaceae</td>\n",
       "      <td>g__Escherichia</td>\n",
       "      <td>s__Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216405</th>\n",
       "      <td>GCF_907275105.1</td>\n",
       "      <td>1269</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>o__Enterobacterales</td>\n",
       "      <td>f__Enterobacteriaceae</td>\n",
       "      <td>g__Escherichia</td>\n",
       "      <td>s__Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216406</th>\n",
       "      <td>GCF_909129205.1</td>\n",
       "      <td>1522</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>o__Enterobacterales</td>\n",
       "      <td>f__Enterobacteriaceae</td>\n",
       "      <td>g__Klebsiella</td>\n",
       "      <td>s__Klebsiella pneumoniae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216407</th>\n",
       "      <td>GCF_909129285.1</td>\n",
       "      <td>1522</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Gammaproteobacteria</td>\n",
       "      <td>o__Enterobacterales</td>\n",
       "      <td>f__Enterobacteriaceae</td>\n",
       "      <td>g__Klebsiella</td>\n",
       "      <td>s__Klebsiella pneumoniae</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>216408 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       assembly_accession  rna_length       kindom             phylum  \\\n",
       "0         GCF_000003135.1        1535  k__Bacteria  p__Actinobacteria   \n",
       "1         GCF_000003215.1        1503  k__Bacteria      p__Firmicutes   \n",
       "2         GCF_000003645.1         849  k__Bacteria      p__Firmicutes   \n",
       "3         GCF_000003925.1        1552  k__Bacteria      p__Firmicutes   \n",
       "4         GCF_000003955.1        1643  k__Bacteria      p__Firmicutes   \n",
       "...                   ...         ...          ...                ...   \n",
       "216403    GCF_907275085.1          88  k__Bacteria  p__Proteobacteria   \n",
       "216404    GCF_907275095.1         299  k__Bacteria  p__Proteobacteria   \n",
       "216405    GCF_907275105.1        1269  k__Bacteria  p__Proteobacteria   \n",
       "216406    GCF_909129205.1        1522  k__Bacteria  p__Proteobacteria   \n",
       "216407    GCF_909129285.1        1522  k__Bacteria  p__Proteobacteria   \n",
       "\n",
       "                         class                 order  \\\n",
       "0             c__Actinomycetia  o__Bifidobacteriales   \n",
       "1                c__Clostridia      o__Eubacteriales   \n",
       "2                   c__Bacilli         o__Bacillales   \n",
       "3                   c__Bacilli         o__Bacillales   \n",
       "4                   c__Bacilli         o__Bacillales   \n",
       "...                        ...                   ...   \n",
       "216403  c__Gammaproteobacteria   o__Enterobacterales   \n",
       "216404  c__Gammaproteobacteria   o__Enterobacterales   \n",
       "216405  c__Gammaproteobacteria   o__Enterobacterales   \n",
       "216406  c__Gammaproteobacteria   o__Enterobacterales   \n",
       "216407  c__Gammaproteobacteria   o__Enterobacterales   \n",
       "\n",
       "                          family               genus  \\\n",
       "0          f__Bifidobacteriaceae  g__Bifidobacterium   \n",
       "1       f__Peptostreptococcaceae   g__Clostridioides   \n",
       "2                 f__Bacillaceae         g__Bacillus   \n",
       "3                 f__Bacillaceae         g__Bacillus   \n",
       "4                 f__Bacillaceae         g__Bacillus   \n",
       "...                          ...                 ...   \n",
       "216403     f__Enterobacteriaceae       g__Klebsiella   \n",
       "216404     f__Enterobacteriaceae      g__Escherichia   \n",
       "216405     f__Enterobacteriaceae      g__Escherichia   \n",
       "216406     f__Enterobacteriaceae       g__Klebsiella   \n",
       "216407     f__Enterobacteriaceae       g__Klebsiella   \n",
       "\n",
       "                            species  \n",
       "0         s__Bifidobacterium longum  \n",
       "1       s__Clostridioides difficile  \n",
       "2                s__Bacillus cereus  \n",
       "3              s__Bacillus mycoides  \n",
       "4                s__Bacillus cereus  \n",
       "...                             ...  \n",
       "216403     s__Klebsiella pneumoniae  \n",
       "216404          s__Escherichia coli  \n",
       "216405          s__Escherichia coli  \n",
       "216406     s__Klebsiella pneumoniae  \n",
       "216407     s__Klebsiella pneumoniae  \n",
       "\n",
       "[216408 rows x 9 columns]"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lineage_info_df = pd.concat([lineage_df[[\"assembly_accession\", \"rna_length\"]], lineage_df.iloc[:, -7:]], axis=1)\n",
    "lineage_info_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "56f4a577",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly_accession</th>\n",
       "      <th>CheckM completeness</th>\n",
       "      <th>CheckM contamination</th>\n",
       "      <th>16S_ID</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>Sum</th>\n",
       "      <th>PKSI</th>\n",
       "      <th>PKSother</th>\n",
       "      <th>NRPS</th>\n",
       "      <th>RiPPs</th>\n",
       "      <th>...</th>\n",
       "      <th>PKS-NRP_Hybrids</th>\n",
       "      <th>Others</th>\n",
       "      <th>rna_length</th>\n",
       "      <th>kindom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_000022305.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.67</td>\n",
       "      <td>refseq4905</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1529</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Betaproteobacteria</td>\n",
       "      <td>o__Burkholderiales</td>\n",
       "      <td>f__Comamonadaceae</td>\n",
       "      <td>g__Acidovorax</td>\n",
       "      <td>s__Acidovorax ebreus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_000025965.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.37</td>\n",
       "      <td>refseq15376</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>1538</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Betaproteobacteria</td>\n",
       "      <td>o__Rhodocyclales</td>\n",
       "      <td>f__Rhodocyclaceae</td>\n",
       "      <td>g__Aromatoleum</td>\n",
       "      <td>s__Aromatoleum aromaticum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_000023105.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.84</td>\n",
       "      <td>refseq45559</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1536</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Actinobacteria</td>\n",
       "      <td>c__Actinomycetia</td>\n",
       "      <td>o__Micrococcales</td>\n",
       "      <td>f__Beutenbergiaceae</td>\n",
       "      <td>g__Beutenbergia</td>\n",
       "      <td>s__Beutenbergia cavernae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_000012365.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.06</td>\n",
       "      <td>refseq128457</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>1532</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Betaproteobacteria</td>\n",
       "      <td>o__Burkholderiales</td>\n",
       "      <td>f__Burkholderiaceae</td>\n",
       "      <td>g__Burkholderia</td>\n",
       "      <td>s__Burkholderia thailandensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_000024165.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.88</td>\n",
       "      <td>refseq35003</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>1538</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Betaproteobacteria</td>\n",
       "      <td>o__unclassified Betaproteobacteria order</td>\n",
       "      <td>f__unclassified Betaproteobacteria family</td>\n",
       "      <td>g__Candidatus Accumulibacter</td>\n",
       "      <td>s__Candidatus Accumulibacter phosphatis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  assembly_accession  CheckM completeness  CheckM contamination        16S_ID  \\\n",
       "0    GCF_000022305.1                100.0                  0.67    refseq4905   \n",
       "1    GCF_000025965.1                100.0                  0.37   refseq15376   \n",
       "2    GCF_000023105.1                100.0                  0.84   refseq45559   \n",
       "3    GCF_000012365.1                100.0                  0.06  refseq128457   \n",
       "4    GCF_000024165.1                100.0                  0.88   refseq35003   \n",
       "\n",
       "    assembly_level  Sum  PKSI  PKSother  NRPS  RiPPs  ...  PKS-NRP_Hybrids  \\\n",
       "0  Complete Genome    3     0         0     0      0  ...                0   \n",
       "1  Complete Genome    7     0         0     0      0  ...                0   \n",
       "2  Complete Genome    1     0         0     0      0  ...                0   \n",
       "3  Complete Genome   20     1         0     5      3  ...                2   \n",
       "4  Complete Genome    8     0         0     2      1  ...                0   \n",
       "\n",
       "   Others  rna_length       kindom             phylum                  class  \\\n",
       "0       2        1529  k__Bacteria  p__Proteobacteria  c__Betaproteobacteria   \n",
       "1       6        1538  k__Bacteria  p__Proteobacteria  c__Betaproteobacteria   \n",
       "2       0        1536  k__Bacteria  p__Actinobacteria       c__Actinomycetia   \n",
       "3       7        1532  k__Bacteria  p__Proteobacteria  c__Betaproteobacteria   \n",
       "4       4        1538  k__Bacteria  p__Proteobacteria  c__Betaproteobacteria   \n",
       "\n",
       "                                      order  \\\n",
       "0                        o__Burkholderiales   \n",
       "1                          o__Rhodocyclales   \n",
       "2                          o__Micrococcales   \n",
       "3                        o__Burkholderiales   \n",
       "4  o__unclassified Betaproteobacteria order   \n",
       "\n",
       "                                      family                         genus  \\\n",
       "0                          f__Comamonadaceae                 g__Acidovorax   \n",
       "1                          f__Rhodocyclaceae                g__Aromatoleum   \n",
       "2                        f__Beutenbergiaceae               g__Beutenbergia   \n",
       "3                        f__Burkholderiaceae               g__Burkholderia   \n",
       "4  f__unclassified Betaproteobacteria family  g__Candidatus Accumulibacter   \n",
       "\n",
       "                                   species  \n",
       "0                     s__Acidovorax ebreus  \n",
       "1                s__Aromatoleum aromaticum  \n",
       "2                 s__Beutenbergia cavernae  \n",
       "3            s__Burkholderia thailandensis  \n",
       "4  s__Candidatus Accumulibacter phosphatis  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df = pd.merge(left=dataset_df, right=lineage_info_df, how=\"inner\", left_on=[\"assembly_accession\"], right_on=[\"assembly_accession\"])\n",
    "merged_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "ff558639",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(175321, 22)"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "69fd9373",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "kindom         1\n",
      "phylum        33\n",
      "class         76\n",
      "order        193\n",
      "family       502\n",
      "genus       2535\n",
      "species    22806\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "unique_counts = merged_df.iloc[:, -7:].nunique()\n",
    "print(unique_counts)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "1b452177",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "phylum\n",
      "p__Proteobacteria                  1119\n",
      "p__Firmicutes                       466\n",
      "p__Actinobacteria                   337\n",
      "p__Bacteroidetes                    297\n",
      "p__Cyanobacteria                     89\n",
      "p__Planctomycetes                    51\n",
      "p__Verrucomicrobia                   22\n",
      "p__Chloroflexi                       21\n",
      "p__Spirochaetes                      16\n",
      "p__Acidobacteria                     16\n",
      "p__Thermotogae                       12\n",
      "p__Aquificae                         12\n",
      "p__Tenericutes                       11\n",
      "p__Synergistetes                     10\n",
      "p__Fusobacteria                      10\n",
      "p__Chlamydiae                         9\n",
      "p__Deinococcus-Thermus                7\n",
      "p__Deferribacteres                    6\n",
      "p__Thermodesulfobacteria              5\n",
      "p__Chlorobi                           4\n",
      "p__unclassified Bacteria              2\n",
      "p__Gemmatimonadetes                   2\n",
      "p__Ignavibacteriae                    2\n",
      "p__Armatimonadetes                    2\n",
      "p__Abditibacteriota                   1\n",
      "p__Nitrospirae                        1\n",
      "p__Nitrospinae                        1\n",
      "p__Lentisphaerae                      1\n",
      "p__Kiritimatiellaeota                 1\n",
      "p__Fibrobacteres                      1\n",
      "p__Dictyoglomi                        1\n",
      "p__Coprothermobacterota               1\n",
      "p__unclassified Bacteria phylum       1\n",
      "Name: genus, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "phylum_genus_counts = merged_df.groupby(\"phylum\")[\"genus\"].nunique().sort_values(ascending=False)\n",
    "print(phylum_genus_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "f72fb1db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(2537)"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylum_genus_counts.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "e697cb0a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "有 2 个 genus 出现在多个 phylum 中：\n",
      "genus\n",
      "g__Ponticoccus                    2\n",
      "g__unclassified Bacteria genus    2\n",
      "Name: phylum, dtype: int64\n",
      "\n",
      "Genus: g__Ponticoccus\n",
      "Phyla: ['p__Proteobacteria' 'p__Actinobacteria']\n",
      "\n",
      "Genus: g__unclassified Bacteria genus\n",
      "Phyla: ['p__unclassified Bacteria phylum' 'p__unclassified Bacteria']\n"
     ]
    }
   ],
   "source": [
    "# 找出在多个 phylum 下出现的 genus\n",
    "cross_phylum_genus = merged_df.groupby(\"genus\")[\"phylum\"].nunique()\n",
    "cross_phylum_genus = cross_phylum_genus[cross_phylum_genus > 1]\n",
    "\n",
    "# 输出总共有多少个 genus 出现在多个 phylum 中\n",
    "print(f\"有 {len(cross_phylum_genus)} 个 genus 出现在多个 phylum 中：\")\n",
    "print(cross_phylum_genus)\n",
    "\n",
    "# 查看这些 genus 分别在哪些 phylum 中\n",
    "for genus in cross_phylum_genus.index:\n",
    "    phyla = merged_df[merged_df[\"genus\"] == genus][\"phylum\"].unique()\n",
    "    print(f\"\\nGenus: {genus}\\nPhyla: {phyla}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4453c97f",
   "metadata": {},
   "source": [
    "- 对于g__Ponticoccus，确实有2个来源'p__Proteobacteria' 'p__Actinobacteria'；\n",
    "- 对于g__unclassified Bacteria genus，考虑将'p__unclassified Bacteria phylum' 'p__unclassified Bacteria'合并"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "3c033222",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly_accession</th>\n",
       "      <th>CheckM completeness</th>\n",
       "      <th>CheckM contamination</th>\n",
       "      <th>16S_ID</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>Sum</th>\n",
       "      <th>PKSI</th>\n",
       "      <th>PKSother</th>\n",
       "      <th>NRPS</th>\n",
       "      <th>RiPPs</th>\n",
       "      <th>...</th>\n",
       "      <th>PKS-NRP_Hybrids</th>\n",
       "      <th>Others</th>\n",
       "      <th>rna_length</th>\n",
       "      <th>kindom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2240</th>\n",
       "      <td>GCF_016097405.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2.25</td>\n",
       "      <td>refseq184069</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1678</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__unclassified Bacteria phylum</td>\n",
       "      <td>c__unclassified Bacteria class</td>\n",
       "      <td>o__unclassified Bacteria order</td>\n",
       "      <td>f__unclassified Bacteria family</td>\n",
       "      <td>g__unclassified Bacteria genus</td>\n",
       "      <td>s__methane-oxidizing endosymbiont of Gigantope...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2252</th>\n",
       "      <td>GCF_016097415.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.72</td>\n",
       "      <td>refseq184073</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>1546</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__unclassified Bacteria phylum</td>\n",
       "      <td>c__unclassified Bacteria class</td>\n",
       "      <td>o__unclassified Bacteria order</td>\n",
       "      <td>f__unclassified Bacteria family</td>\n",
       "      <td>g__unclassified Bacteria genus</td>\n",
       "      <td>s__sulfur-oxidizing endosymbiont of Gigantopel...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     assembly_accession  CheckM completeness  CheckM contamination  \\\n",
       "2240    GCF_016097405.1                100.0                  2.25   \n",
       "2252    GCF_016097415.1                100.0                  1.72   \n",
       "\n",
       "            16S_ID assembly_level  Sum  PKSI  PKSother  NRPS  RiPPs  ...  \\\n",
       "2240  refseq184069       Scaffold    3     0         0     0      1  ...   \n",
       "2252  refseq184073       Scaffold    8     0         1     1      1  ...   \n",
       "\n",
       "      PKS-NRP_Hybrids  Others  rna_length       kindom  \\\n",
       "2240                0       1        1678  k__Bacteria   \n",
       "2252                0       4        1546  k__Bacteria   \n",
       "\n",
       "                               phylum                           class  \\\n",
       "2240  p__unclassified Bacteria phylum  c__unclassified Bacteria class   \n",
       "2252  p__unclassified Bacteria phylum  c__unclassified Bacteria class   \n",
       "\n",
       "                               order                           family  \\\n",
       "2240  o__unclassified Bacteria order  f__unclassified Bacteria family   \n",
       "2252  o__unclassified Bacteria order  f__unclassified Bacteria family   \n",
       "\n",
       "                               genus  \\\n",
       "2240  g__unclassified Bacteria genus   \n",
       "2252  g__unclassified Bacteria genus   \n",
       "\n",
       "                                                species  \n",
       "2240  s__methane-oxidizing endosymbiont of Gigantope...  \n",
       "2252  s__sulfur-oxidizing endosymbiont of Gigantopel...  \n",
       "\n",
       "[2 rows x 22 columns]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df[merged_df['phylum'] == 'p__unclassified Bacteria phylum']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "1136cb45",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly_accession</th>\n",
       "      <th>CheckM completeness</th>\n",
       "      <th>CheckM contamination</th>\n",
       "      <th>16S_ID</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>Sum</th>\n",
       "      <th>PKSI</th>\n",
       "      <th>PKSother</th>\n",
       "      <th>NRPS</th>\n",
       "      <th>RiPPs</th>\n",
       "      <th>...</th>\n",
       "      <th>PKS-NRP_Hybrids</th>\n",
       "      <th>Others</th>\n",
       "      <th>rna_length</th>\n",
       "      <th>kindom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>122157</th>\n",
       "      <td>GCF_900128535.1</td>\n",
       "      <td>98.28</td>\n",
       "      <td>3.45</td>\n",
       "      <td>refseq166441</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1540</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__unclassified Bacteria</td>\n",
       "      <td>phylum   c__unclassified Bacteria class</td>\n",
       "      <td>o__unclassified Bacteria order</td>\n",
       "      <td>f__unclassified Bacteria family</td>\n",
       "      <td>g__unclassified Bacteria genus</td>\n",
       "      <td>s__bacterium endosymbiont of Bathymodiolus sp....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152341</th>\n",
       "      <td>GCF_900128525.1</td>\n",
       "      <td>97.19</td>\n",
       "      <td>1.97</td>\n",
       "      <td>refseq152190</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1539</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__unclassified Bacteria</td>\n",
       "      <td>phylum   c__unclassified Bacteria class</td>\n",
       "      <td>o__unclassified Bacteria order</td>\n",
       "      <td>f__unclassified Bacteria family</td>\n",
       "      <td>g__unclassified Bacteria genus</td>\n",
       "      <td>s__methanotrophic endosymbiont of Bathymodiolu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159664</th>\n",
       "      <td>GCF_900128515.1</td>\n",
       "      <td>96.55</td>\n",
       "      <td>3.45</td>\n",
       "      <td>refseq166440</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1549</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__unclassified Bacteria</td>\n",
       "      <td>phylum   c__unclassified Bacteria class</td>\n",
       "      <td>o__unclassified Bacteria order</td>\n",
       "      <td>f__unclassified Bacteria family</td>\n",
       "      <td>g__unclassified Bacteria genus</td>\n",
       "      <td>s__Bathymodiolus heckerae thiotrophic gill sym...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161989</th>\n",
       "      <td>GCF_009910325.1</td>\n",
       "      <td>96.21</td>\n",
       "      <td>3.98</td>\n",
       "      <td>refseq57595</td>\n",
       "      <td>Contig</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1511</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__unclassified Bacteria</td>\n",
       "      <td>phylum   c__unclassified Bacteria class</td>\n",
       "      <td>o__unclassified Bacteria order</td>\n",
       "      <td>f__unclassified Bacteria family</td>\n",
       "      <td>g__Isachenkonia</td>\n",
       "      <td>s__Isachenkonia alkalipeptolytica</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171756</th>\n",
       "      <td>GCF_903642095.1</td>\n",
       "      <td>93.64</td>\n",
       "      <td>0.56</td>\n",
       "      <td>refseq105622</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1534</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__unclassified Bacteria</td>\n",
       "      <td>phylum   c__unclassified Bacteria class</td>\n",
       "      <td>o__unclassified Bacteria order</td>\n",
       "      <td>f__unclassified Bacteria family</td>\n",
       "      <td>g__unclassified Bacteria genus</td>\n",
       "      <td>s__Bathymodiolus brooksi thiotrophic gill symb...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171757</th>\n",
       "      <td>GCF_903642125.1</td>\n",
       "      <td>93.64</td>\n",
       "      <td>0.67</td>\n",
       "      <td>refseq105623</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1534</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__unclassified Bacteria</td>\n",
       "      <td>phylum   c__unclassified Bacteria class</td>\n",
       "      <td>o__unclassified Bacteria order</td>\n",
       "      <td>f__unclassified Bacteria family</td>\n",
       "      <td>g__unclassified Bacteria genus</td>\n",
       "      <td>s__Bathymodiolus brooksi thiotrophic gill symb...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172156</th>\n",
       "      <td>GCF_016592615.1</td>\n",
       "      <td>93.41</td>\n",
       "      <td>0.75</td>\n",
       "      <td>refseq184977</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1545</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__unclassified Bacteria</td>\n",
       "      <td>phylum   c__unclassified Bacteria class</td>\n",
       "      <td>o__unclassified Bacteria order</td>\n",
       "      <td>f__unclassified Bacteria family</td>\n",
       "      <td>g__unclassified Bacteria genus</td>\n",
       "      <td>s__Isorropodon fossajaponicum symbiont</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172256</th>\n",
       "      <td>GCF_900127505.1</td>\n",
       "      <td>93.36</td>\n",
       "      <td>1.40</td>\n",
       "      <td>refseq151751</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1540</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__unclassified Bacteria</td>\n",
       "      <td>phylum   c__unclassified Bacteria class</td>\n",
       "      <td>o__unclassified Bacteria order</td>\n",
       "      <td>f__unclassified Bacteria family</td>\n",
       "      <td>g__unclassified Bacteria genus</td>\n",
       "      <td>s__thiotrophic endosymbiont of Bathymodiolus p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173985</th>\n",
       "      <td>GCF_016592595.1</td>\n",
       "      <td>91.91</td>\n",
       "      <td>0.56</td>\n",
       "      <td>refseq184976</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1546</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__unclassified Bacteria</td>\n",
       "      <td>phylum   c__unclassified Bacteria class</td>\n",
       "      <td>o__unclassified Bacteria order</td>\n",
       "      <td>f__unclassified Bacteria family</td>\n",
       "      <td>g__unclassified Bacteria genus</td>\n",
       "      <td>s__Abyssogena phaseoliformis symbiont</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       assembly_accession  CheckM completeness  CheckM contamination  \\\n",
       "122157    GCF_900128535.1                98.28                  3.45   \n",
       "152341    GCF_900128525.1                97.19                  1.97   \n",
       "159664    GCF_900128515.1                96.55                  3.45   \n",
       "161989    GCF_009910325.1                96.21                  3.98   \n",
       "171756    GCF_903642095.1                93.64                  0.56   \n",
       "171757    GCF_903642125.1                93.64                  0.67   \n",
       "172156    GCF_016592615.1                93.41                  0.75   \n",
       "172256    GCF_900127505.1                93.36                  1.40   \n",
       "173985    GCF_016592595.1                91.91                  0.56   \n",
       "\n",
       "              16S_ID   assembly_level  Sum  PKSI  PKSother  NRPS  RiPPs  ...  \\\n",
       "122157  refseq166441         Scaffold    3     0         0     0      1  ...   \n",
       "152341  refseq152190         Scaffold    4     0         0     1      1  ...   \n",
       "159664  refseq166440         Scaffold    3     0         0     0      1  ...   \n",
       "161989   refseq57595           Contig    1     0         0     0      1  ...   \n",
       "171756  refseq105622         Scaffold    0     0         0     0      0  ...   \n",
       "171757  refseq105623         Scaffold    0     0         0     0      0  ...   \n",
       "172156  refseq184977  Complete Genome    3     0         0     0      1  ...   \n",
       "172256  refseq151751         Scaffold    1     0         0     0      1  ...   \n",
       "173985  refseq184976  Complete Genome    2     0         0     0      1  ...   \n",
       "\n",
       "        PKS-NRP_Hybrids  Others  rna_length       kindom  \\\n",
       "122157                0       2        1540  k__Bacteria   \n",
       "152341                0       1        1539  k__Bacteria   \n",
       "159664                0       2        1549  k__Bacteria   \n",
       "161989                0       0        1511  k__Bacteria   \n",
       "171756                0       0        1534  k__Bacteria   \n",
       "171757                0       0        1534  k__Bacteria   \n",
       "172156                0       2        1545  k__Bacteria   \n",
       "172256                0       0        1540  k__Bacteria   \n",
       "173985                0       1        1546  k__Bacteria   \n",
       "\n",
       "                          phylum                                    class  \\\n",
       "122157  p__unclassified Bacteria  phylum   c__unclassified Bacteria class   \n",
       "152341  p__unclassified Bacteria  phylum   c__unclassified Bacteria class   \n",
       "159664  p__unclassified Bacteria  phylum   c__unclassified Bacteria class   \n",
       "161989  p__unclassified Bacteria  phylum   c__unclassified Bacteria class   \n",
       "171756  p__unclassified Bacteria  phylum   c__unclassified Bacteria class   \n",
       "171757  p__unclassified Bacteria  phylum   c__unclassified Bacteria class   \n",
       "172156  p__unclassified Bacteria  phylum   c__unclassified Bacteria class   \n",
       "172256  p__unclassified Bacteria  phylum   c__unclassified Bacteria class   \n",
       "173985  p__unclassified Bacteria  phylum   c__unclassified Bacteria class   \n",
       "\n",
       "                                 order                           family  \\\n",
       "122157  o__unclassified Bacteria order  f__unclassified Bacteria family   \n",
       "152341  o__unclassified Bacteria order  f__unclassified Bacteria family   \n",
       "159664  o__unclassified Bacteria order  f__unclassified Bacteria family   \n",
       "161989  o__unclassified Bacteria order  f__unclassified Bacteria family   \n",
       "171756  o__unclassified Bacteria order  f__unclassified Bacteria family   \n",
       "171757  o__unclassified Bacteria order  f__unclassified Bacteria family   \n",
       "172156  o__unclassified Bacteria order  f__unclassified Bacteria family   \n",
       "172256  o__unclassified Bacteria order  f__unclassified Bacteria family   \n",
       "173985  o__unclassified Bacteria order  f__unclassified Bacteria family   \n",
       "\n",
       "                                 genus  \\\n",
       "122157  g__unclassified Bacteria genus   \n",
       "152341  g__unclassified Bacteria genus   \n",
       "159664  g__unclassified Bacteria genus   \n",
       "161989                 g__Isachenkonia   \n",
       "171756  g__unclassified Bacteria genus   \n",
       "171757  g__unclassified Bacteria genus   \n",
       "172156  g__unclassified Bacteria genus   \n",
       "172256  g__unclassified Bacteria genus   \n",
       "173985  g__unclassified Bacteria genus   \n",
       "\n",
       "                                                  species  \n",
       "122157  s__bacterium endosymbiont of Bathymodiolus sp....  \n",
       "152341  s__methanotrophic endosymbiont of Bathymodiolu...  \n",
       "159664  s__Bathymodiolus heckerae thiotrophic gill sym...  \n",
       "161989                  s__Isachenkonia alkalipeptolytica  \n",
       "171756  s__Bathymodiolus brooksi thiotrophic gill symb...  \n",
       "171757  s__Bathymodiolus brooksi thiotrophic gill symb...  \n",
       "172156             s__Isorropodon fossajaponicum symbiont  \n",
       "172256  s__thiotrophic endosymbiont of Bathymodiolus p...  \n",
       "173985              s__Abyssogena phaseoliformis symbiont  \n",
       "\n",
       "[9 rows x 22 columns]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df[merged_df['phylum'] == 'p__unclassified Bacteria']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "35f0ac77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly_accession</th>\n",
       "      <th>CheckM completeness</th>\n",
       "      <th>CheckM contamination</th>\n",
       "      <th>16S_ID</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>Sum</th>\n",
       "      <th>PKSI</th>\n",
       "      <th>PKSother</th>\n",
       "      <th>NRPS</th>\n",
       "      <th>RiPPs</th>\n",
       "      <th>...</th>\n",
       "      <th>PKS-NRP_Hybrids</th>\n",
       "      <th>Others</th>\n",
       "      <th>rna_length</th>\n",
       "      <th>kindom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [assembly_accession, CheckM completeness, CheckM contamination, 16S_ID, assembly_level, Sum, PKSI, PKSother, NRPS, RiPPs, Saccharides, Terpene, PKS-NRP_Hybrids, Others, rna_length, kindom, phylum, class, order, family, genus, species]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 22 columns]"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_merged_df = merged_df.copy()\n",
    "new_merged_df.loc[new_merged_df['phylum'] == 'p__unclassified Bacteria phylum', 'phylum'] = 'p__unclassified Bacteria'\n",
    "\n",
    "new_merged_df[new_merged_df['phylum'] == 'p__unclassified Bacteria phylum']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "d111723c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(2536)"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylum_genus_counts = new_merged_df.groupby(\"phylum\")[\"genus\"].nunique().sort_values(ascending=False)\n",
    "phylum_genus_counts.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "98c2f9cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "kindom         1\n",
      "phylum        32\n",
      "class         76\n",
      "order        193\n",
      "family       502\n",
      "genus       2535\n",
      "species    22806\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "unique_counts = new_merged_df.iloc[:, -7:].nunique()\n",
    "print(unique_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "57094af0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "kindom         1\n",
       "phylum        32\n",
       "class         76\n",
       "order        193\n",
       "family       502\n",
       "genus       2536\n",
       "species    22806\n",
       "dtype: int64"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_counts[\"genus\"] += 1  # g__Ponticoccus 算2个不同类\n",
    "unique_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "e25627f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# 创建一个 ExcelWriter 对象\n",
    "with pd.ExcelWriter(\"data/Fig1.xlsx\") as writer:\n",
    "    # 将 unique_counts 保存到 sheet1\n",
    "    unique_counts_df = unique_counts.to_frame(name=\"unique_counts\")  # 转换为 DataFrame\n",
    "    unique_counts_df.to_excel(writer, sheet_name=\"sheet1\")\n",
    "\n",
    "    # 将 phylum_genus_counts 保存到 sheet2\n",
    "    phylum_genus_counts_df = phylum_genus_counts.to_frame(name=\"phylum_genus_counts\")  # 转换为 DataFrame\n",
    "    phylum_genus_counts_df.to_excel(writer, sheet_name=\"sheet2\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "d9e122e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "phylum                    genus                                     \n",
       "p__Abditibacteriota       g__Abditibacterium                             2.0\n",
       "p__Acidobacteria          g__Acidipila                                   7.0\n",
       "                          g__Acidisarcina                               11.0\n",
       "                          g__Acidobacterium                              6.0\n",
       "                          g__Bryobacter                                  1.0\n",
       "                                                                        ... \n",
       "p__Verrucomicrobia        g__Terrimicrobium                              8.0\n",
       "                          g__Verrucomicrobium                            7.0\n",
       "                          g__unclassified Methylacidiphilaceae genus     6.0\n",
       "p__unclassified Bacteria  g__Isachenkonia                                1.0\n",
       "                          g__unclassified Bacteria genus                 2.7\n",
       "Name: Sum, Length: 2536, dtype: float64"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 统计每个phylum-genus下的BGC的平均总数\n",
    "mean_sum_result = new_merged_df.groupby(by=[\"phylum\", \"genus\"])[\"Sum\"].mean()\n",
    "mean_sum_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "b9ad8f04",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "phylum                    genus                                     \n",
       "p__Abditibacteriota       g__Abditibacterium                            0.000000\n",
       "p__Acidobacteria          g__Acidipila                                  0.698970\n",
       "                          g__Acidisarcina                               0.000000\n",
       "                          g__Acidobacterium                             0.477121\n",
       "                          g__Bryobacter                                 0.000000\n",
       "                                                                          ...   \n",
       "p__Verrucomicrobia        g__Terrimicrobium                             0.000000\n",
       "                          g__Verrucomicrobium                           0.301030\n",
       "                          g__unclassified Methylacidiphilaceae genus    0.000000\n",
       "p__unclassified Bacteria  g__Isachenkonia                               0.000000\n",
       "                          g__unclassified Bacteria genus                1.000000\n",
       "Name: size, Length: 2536, dtype: float64"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "# 统计每组的行数\n",
    "group_sizes = new_merged_df.groupby(by=[\"phylum\", \"genus\"]).size()\n",
    "group_sizes.name = \"size\"\n",
    "group_sizes\n",
    "\n",
    "# 对 group_sizes 取 log10\n",
    "group_sizes_log10 = np.log10(group_sizes)\n",
    "\n",
    "group_sizes_log10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "3261d9da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'refseq1'"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts = new_merged_df['16S_ID'].value_counts()\n",
    "max_count = counts.max()\n",
    "most_frequent_count = counts[counts == max_count]\n",
    "most_frequent_count.index[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "17dbb1cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/3c/v18vsqt50jzgnmgzv858z_rm0000gn/T/ipykernel_20195/2650675103.py:18: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  most_frequent_16S_result = new_merged_df.groupby(by=[\"phylum\", \"genus\"]).apply(most_frequent_16S)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "phylum                    genus                                     \n",
       "p__Abditibacteriota       g__Abditibacterium                             refseq52887\n",
       "p__Acidobacteria          g__Acidipila                                   refseq33882\n",
       "                          g__Acidisarcina                                refseq62194\n",
       "                          g__Acidobacterium                             refseq128816\n",
       "                          g__Bryobacter                                  refseq44527\n",
       "                                                                            ...     \n",
       "p__Verrucomicrobia        g__Terrimicrobium                              refseq97457\n",
       "                          g__Verrucomicrobium                           refseq125928\n",
       "                          g__unclassified Methylacidiphilaceae genus    refseq115367\n",
       "p__unclassified Bacteria  g__Isachenkonia                                refseq57595\n",
       "                          g__unclassified Bacteria genus                refseq166440\n",
       "Length: 2536, dtype: object"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 计算每个分组中的 \"16S_ID\" 出现次数\n",
    "def most_frequent_16S(group):\n",
    "    # 计算每个 \"16S_ID\" 的出现次数\n",
    "    counts = group['16S_ID'].value_counts()\n",
    "    \n",
    "    # 获取最大出现次数\n",
    "    max_count = counts.max()\n",
    "    \n",
    "    # 获取所有出现次数为最大值的 16S_ID\n",
    "    most_frequent_count = counts[counts == max_count]\n",
    "    \n",
    "    # 如果只有一个 16S_ID 出现次数最多，返回该 16S_ID；如果有多个，则随机选择一个\n",
    "    most_frequent = most_frequent_count.sample(n=1, random_state=2).index[0]  # 随机选择一个\n",
    "    \n",
    "    return most_frequent\n",
    "\n",
    "# 统计每个phylum-genus下的代表性16S\n",
    "most_frequent_16S_result = new_merged_df.groupby(by=[\"phylum\", \"genus\"]).apply(most_frequent_16S)\n",
    "most_frequent_16S_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "11f89dc0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "phylum             genus                                  \n",
      "p__Actinobacteria  g__Cryocola                                 refseq7830\n",
      "                   g__Leifsonia                                refseq7830\n",
      "p__Bacteroidetes   g__Fluviicola                              refseq47137\n",
      "                   g__unclassified Crocinitomicaceae genus    refseq47137\n",
      "p__Proteobacteria  g__Aggregicoccus                           refseq45025\n",
      "                   g__Simulacricoccus                         refseq45025\n",
      "dtype: object\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "6"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 查找重复的值\n",
    "duplicates = most_frequent_16S_result[most_frequent_16S_result.duplicated(keep=False)]\n",
    "\n",
    "# 打印重复的值\n",
    "print(duplicates)\n",
    "len(duplicates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "9b43b44c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>Sum</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>phylum</th>\n",
       "      <th>genus</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p__Abditibacteriota</th>\n",
       "      <th>g__Abditibacterium</th>\n",
       "      <td>refseq52887</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">p__Acidobacteria</th>\n",
       "      <th>g__Acidipila</th>\n",
       "      <td>refseq33882</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.698970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Acidisarcina</th>\n",
       "      <td>refseq62194</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Acidobacterium</th>\n",
       "      <td>refseq128816</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.477121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Bryobacter</th>\n",
       "      <td>refseq44527</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">p__Verrucomicrobia</th>\n",
       "      <th>g__Terrimicrobium</th>\n",
       "      <td>refseq97457</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Verrucomicrobium</th>\n",
       "      <td>refseq125928</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.301030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__unclassified Methylacidiphilaceae genus</th>\n",
       "      <td>refseq115367</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">p__unclassified Bacteria</th>\n",
       "      <th>g__Isachenkonia</th>\n",
       "      <td>refseq57595</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__unclassified Bacteria genus</th>\n",
       "      <td>refseq166440</td>\n",
       "      <td>2.7</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2536 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                0  \\\n",
       "phylum                   genus                                                      \n",
       "p__Abditibacteriota      g__Abditibacterium                           refseq52887   \n",
       "p__Acidobacteria         g__Acidipila                                 refseq33882   \n",
       "                         g__Acidisarcina                              refseq62194   \n",
       "                         g__Acidobacterium                           refseq128816   \n",
       "                         g__Bryobacter                                refseq44527   \n",
       "...                                                                           ...   \n",
       "p__Verrucomicrobia       g__Terrimicrobium                            refseq97457   \n",
       "                         g__Verrucomicrobium                         refseq125928   \n",
       "                         g__unclassified Methylacidiphilaceae genus  refseq115367   \n",
       "p__unclassified Bacteria g__Isachenkonia                              refseq57595   \n",
       "                         g__unclassified Bacteria genus              refseq166440   \n",
       "\n",
       "                                                                      Sum  \\\n",
       "phylum                   genus                                              \n",
       "p__Abditibacteriota      g__Abditibacterium                           2.0   \n",
       "p__Acidobacteria         g__Acidipila                                 7.0   \n",
       "                         g__Acidisarcina                             11.0   \n",
       "                         g__Acidobacterium                            6.0   \n",
       "                         g__Bryobacter                                1.0   \n",
       "...                                                                   ...   \n",
       "p__Verrucomicrobia       g__Terrimicrobium                            8.0   \n",
       "                         g__Verrucomicrobium                          7.0   \n",
       "                         g__unclassified Methylacidiphilaceae genus   6.0   \n",
       "p__unclassified Bacteria g__Isachenkonia                              1.0   \n",
       "                         g__unclassified Bacteria genus               2.7   \n",
       "\n",
       "                                                                         size  \n",
       "phylum                   genus                                                 \n",
       "p__Abditibacteriota      g__Abditibacterium                          0.000000  \n",
       "p__Acidobacteria         g__Acidipila                                0.698970  \n",
       "                         g__Acidisarcina                             0.000000  \n",
       "                         g__Acidobacterium                           0.477121  \n",
       "                         g__Bryobacter                               0.000000  \n",
       "...                                                                       ...  \n",
       "p__Verrucomicrobia       g__Terrimicrobium                           0.000000  \n",
       "                         g__Verrucomicrobium                         0.301030  \n",
       "                         g__unclassified Methylacidiphilaceae genus  0.000000  \n",
       "p__unclassified Bacteria g__Isachenkonia                             0.000000  \n",
       "                         g__unclassified Bacteria genus              1.000000  \n",
       "\n",
       "[2536 rows x 3 columns]"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 3. 合并两个结果\n",
    "merged_result = pd.concat([most_frequent_16S_result, mean_sum_result, group_sizes_log10], axis=1)\n",
    "merged_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "b0c30b55",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_result.to_csv(\"data/Fig1C.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "c36a0500",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['refseq52887',\n",
       " 'refseq33882',\n",
       " 'refseq62194',\n",
       " 'refseq128816',\n",
       " 'refseq44527',\n",
       " 'refseq165003',\n",
       " 'refseq128507',\n",
       " 'refseq133496',\n",
       " 'refseq116521',\n",
       " 'refseq25627',\n",
       " 'refseq118006',\n",
       " 'refseq120225',\n",
       " 'refseq200266',\n",
       " 'refseq131358',\n",
       " 'refseq178884',\n",
       " 'refseq95159',\n",
       " 'refseq31098',\n",
       " 'refseq43207',\n",
       " 'refseq7277',\n",
       " 'refseq87522',\n",
       " 'refseq11747',\n",
       " 'refseq17767',\n",
       " 'refseq116177',\n",
       " 'refseq12482',\n",
       " 'refseq52839',\n",
       " 'refseq56577',\n",
       " 'refseq27374',\n",
       " 'refseq58844',\n",
       " 'refseq102276',\n",
       " 'refseq30509',\n",
       " 'refseq25818',\n",
       " 'refseq142014',\n",
       " 'refseq70341',\n",
       " 'refseq149173',\n",
       " 'refseq156277',\n",
       " 'refseq60893',\n",
       " 'refseq127199',\n",
       " 'refseq32978',\n",
       " 'refseq19813',\n",
       " 'refseq182367',\n",
       " 'refseq44142',\n",
       " 'refseq55740',\n",
       " 'refseq153899',\n",
       " 'refseq134430',\n",
       " 'refseq22917',\n",
       " 'refseq14480',\n",
       " 'refseq115947',\n",
       " 'refseq5057',\n",
       " 'refseq10441',\n",
       " 'refseq37712',\n",
       " 'refseq116074',\n",
       " 'refseq4244',\n",
       " 'refseq94565',\n",
       " 'refseq67443',\n",
       " 'refseq115951',\n",
       " 'refseq58838',\n",
       " 'refseq104511',\n",
       " 'refseq7740',\n",
       " 'refseq2106',\n",
       " 'refseq56757',\n",
       " 'refseq12641',\n",
       " 'refseq185999',\n",
       " 'refseq8591',\n",
       " 'refseq56932',\n",
       " 'refseq175510',\n",
       " 'refseq43223',\n",
       " 'refseq45559',\n",
       " 'refseq318',\n",
       " 'refseq165233',\n",
       " 'refseq190066',\n",
       " 'refseq50500',\n",
       " 'refseq135621',\n",
       " 'refseq38197',\n",
       " 'refseq7267',\n",
       " 'refseq108161',\n",
       " 'refseq154950',\n",
       " 'refseq38403',\n",
       " 'refseq59344',\n",
       " 'refseq44314',\n",
       " 'refseq29783',\n",
       " 'refseq22232',\n",
       " 'refseq81367',\n",
       " 'refseq24089',\n",
       " 'refseq130407',\n",
       " 'refseq15470',\n",
       " 'refseq14108',\n",
       " 'refseq180543',\n",
       " 'refseq44585',\n",
       " 'refseq2256',\n",
       " 'refseq37880',\n",
       " 'refseq63443',\n",
       " 'refseq115914',\n",
       " 'refseq199277',\n",
       " 'refseq47566',\n",
       " 'refseq594',\n",
       " 'refseq203569',\n",
       " 'refseq11408',\n",
       " 'refseq9611',\n",
       " 'refseq7830',\n",
       " 'refseq28468',\n",
       " 'refseq151889',\n",
       " 'refseq21339',\n",
       " 'refseq371',\n",
       " 'refseq165101',\n",
       " 'refseq36539',\n",
       " 'refseq35918',\n",
       " 'refseq34738',\n",
       " 'refseq7048',\n",
       " 'refseq2061',\n",
       " 'refseq62487',\n",
       " 'refseq35114',\n",
       " 'refseq173653',\n",
       " 'refseq44009',\n",
       " 'refseq11609',\n",
       " 'refseq4796',\n",
       " 'refseq69919',\n",
       " 'refseq20790',\n",
       " 'refseq79761',\n",
       " 'refseq17717',\n",
       " 'refseq35318',\n",
       " 'refseq50022',\n",
       " 'refseq89266',\n",
       " 'refseq116757',\n",
       " 'refseq199768',\n",
       " 'refseq45639',\n",
       " 'refseq36161',\n",
       " 'refseq166543',\n",
       " 'refseq15568',\n",
       " 'refseq31451',\n",
       " 'refseq49305',\n",
       " 'refseq40071',\n",
       " 'refseq5941',\n",
       " 'refseq30343',\n",
       " 'refseq62586',\n",
       " 'refseq53270',\n",
       " 'refseq177990',\n",
       " 'refseq179834',\n",
       " 'refseq54745',\n",
       " 'refseq2380',\n",
       " 'refseq23939',\n",
       " 'refseq68161',\n",
       " 'refseq54456',\n",
       " 'refseq60477',\n",
       " 'refseq117396',\n",
       " 'refseq5953',\n",
       " 'refseq130386',\n",
       " 'refseq56045',\n",
       " 'refseq4724',\n",
       " 'refseq51723',\n",
       " 'refseq35174',\n",
       " 'refseq29287',\n",
       " 'refseq31166',\n",
       " 'refseq82351',\n",
       " 'refseq115954',\n",
       " 'refseq41996',\n",
       " 'refseq168249',\n",
       " 'refseq14053',\n",
       " 'refseq76733',\n",
       " 'refseq128159',\n",
       " 'refseq94128',\n",
       " 'refseq37476',\n",
       " 'refseq116176',\n",
       " 'refseq48620',\n",
       " 'refseq40175',\n",
       " 'refseq50337',\n",
       " 'refseq55731',\n",
       " 'refseq145887',\n",
       " 'refseq160968',\n",
       " 'refseq26333',\n",
       " 'refseq145601',\n",
       " 'refseq36299',\n",
       " 'refseq19956',\n",
       " 'refseq10448',\n",
       " 'refseq130140',\n",
       " 'refseq181053',\n",
       " 'refseq57581',\n",
       " 'refseq25494',\n",
       " 'refseq51940',\n",
       " 'refseq14286',\n",
       " 'refseq150868',\n",
       " 'refseq28670',\n",
       " 'refseq35021',\n",
       " 'refseq4072',\n",
       " 'refseq24940',\n",
       " 'refseq167567',\n",
       " 'refseq46621',\n",
       " 'refseq96466',\n",
       " 'refseq160088',\n",
       " 'refseq24512',\n",
       " 'refseq70337',\n",
       " 'refseq28467',\n",
       " 'refseq55825',\n",
       " 'refseq30000',\n",
       " 'refseq7830',\n",
       " 'refseq151451',\n",
       " 'refseq21749',\n",
       " 'refseq180022',\n",
       " 'refseq14093',\n",
       " 'refseq41939',\n",
       " 'refseq38400',\n",
       " 'refseq180173',\n",
       " 'refseq180029',\n",
       " 'refseq32622',\n",
       " 'refseq167628',\n",
       " 'refseq28564',\n",
       " 'refseq58058',\n",
       " 'refseq40611',\n",
       " 'refseq2389',\n",
       " 'refseq41768',\n",
       " 'refseq5593',\n",
       " 'refseq164811',\n",
       " 'refseq14298',\n",
       " 'refseq149023',\n",
       " 'refseq23847',\n",
       " 'refseq23911',\n",
       " 'refseq5153',\n",
       " 'refseq47642',\n",
       " 'refseq54878',\n",
       " 'refseq18046',\n",
       " 'refseq47050',\n",
       " 'refseq42450',\n",
       " 'refseq2',\n",
       " 'refseq55',\n",
       " 'refseq115837',\n",
       " 'refseq26487',\n",
       " 'refseq2730',\n",
       " 'refseq54453',\n",
       " 'refseq41964',\n",
       " 'refseq70435',\n",
       " 'refseq17481',\n",
       " 'refseq46091',\n",
       " 'refseq119045',\n",
       " 'refseq10654',\n",
       " 'refseq88051',\n",
       " 'refseq2223',\n",
       " 'refseq34191',\n",
       " 'refseq3869',\n",
       " 'refseq125961',\n",
       " 'refseq104969',\n",
       " 'refseq22086',\n",
       " 'refseq151875',\n",
       " 'refseq119475',\n",
       " 'refseq43902',\n",
       " 'refseq20009',\n",
       " 'refseq53834',\n",
       " 'refseq9469',\n",
       " 'refseq58634',\n",
       " 'refseq144760',\n",
       " 'refseq22842',\n",
       " 'refseq24232',\n",
       " 'refseq166930',\n",
       " 'refseq127467',\n",
       " 'refseq177963',\n",
       " 'refseq19827',\n",
       " 'refseq45968',\n",
       " 'refseq57883',\n",
       " 'refseq116655',\n",
       " 'refseq21596',\n",
       " 'refseq199688',\n",
       " 'refseq41955',\n",
       " 'refseq115958',\n",
       " 'refseq5032',\n",
       " 'refseq70326',\n",
       " 'refseq12716',\n",
       " 'refseq159784',\n",
       " 'refseq1430',\n",
       " 'refseq145657',\n",
       " 'refseq32853',\n",
       " 'refseq146902',\n",
       " 'refseq34805',\n",
       " 'refseq116081',\n",
       " 'refseq81455',\n",
       " 'refseq16563',\n",
       " 'refseq127883',\n",
       " 'refseq142776',\n",
       " 'refseq155540',\n",
       " 'refseq4272',\n",
       " 'refseq32390',\n",
       " 'refseq184641',\n",
       " 'refseq44428',\n",
       " 'refseq57214',\n",
       " 'refseq151970',\n",
       " 'refseq164059',\n",
       " 'refseq103394',\n",
       " 'refseq1732',\n",
       " 'refseq15336',\n",
       " 'refseq106871',\n",
       " 'refseq900',\n",
       " 'refseq36974',\n",
       " 'refseq100375',\n",
       " 'refseq4294',\n",
       " 'refseq127539',\n",
       " 'refseq196650',\n",
       " 'refseq38991',\n",
       " 'refseq17272',\n",
       " 'refseq12219',\n",
       " 'refseq123985',\n",
       " 'refseq26592',\n",
       " 'refseq1641',\n",
       " 'refseq202589',\n",
       " 'refseq13515',\n",
       " 'refseq18126',\n",
       " 'refseq55258',\n",
       " 'refseq56747',\n",
       " 'refseq186010',\n",
       " 'refseq118300',\n",
       " 'refseq120038',\n",
       " 'refseq26683',\n",
       " 'refseq56191',\n",
       " 'refseq28362',\n",
       " 'refseq115792',\n",
       " 'refseq62453',\n",
       " 'refseq26056',\n",
       " 'refseq127969',\n",
       " 'refseq107046',\n",
       " 'refseq130435',\n",
       " 'refseq42801',\n",
       " 'refseq7708',\n",
       " 'refseq153072',\n",
       " 'refseq21284',\n",
       " 'refseq1415',\n",
       " 'refseq151254',\n",
       " 'refseq21420',\n",
       " 'refseq89690',\n",
       " 'refseq32630',\n",
       " 'refseq106737',\n",
       " 'refseq102923',\n",
       " 'refseq48319',\n",
       " 'refseq91179',\n",
       " 'refseq44487',\n",
       " 'refseq115812',\n",
       " 'refseq27331',\n",
       " 'refseq153041',\n",
       " 'refseq46385',\n",
       " 'refseq123417',\n",
       " 'refseq60240',\n",
       " 'refseq77821',\n",
       " 'refseq4533',\n",
       " 'refseq26830',\n",
       " 'refseq106165',\n",
       " 'refseq64717',\n",
       " 'refseq139267',\n",
       " 'refseq180033',\n",
       " 'refseq19371',\n",
       " 'refseq25462',\n",
       " 'refseq24248',\n",
       " 'refseq30148',\n",
       " 'refseq53077',\n",
       " 'refseq127514',\n",
       " 'refseq110301',\n",
       " 'refseq47713',\n",
       " 'refseq116315',\n",
       " 'refseq26275',\n",
       " 'refseq107067',\n",
       " 'refseq42579',\n",
       " 'refseq115217',\n",
       " 'refseq152092',\n",
       " 'refseq144453',\n",
       " 'refseq117508',\n",
       " 'refseq22952',\n",
       " 'refseq144543',\n",
       " 'refseq47838',\n",
       " 'refseq116157',\n",
       " 'refseq22285',\n",
       " 'refseq130211',\n",
       " 'refseq144413',\n",
       " 'refseq64471',\n",
       " 'refseq46779',\n",
       " 'refseq23538',\n",
       " 'refseq37662',\n",
       " 'refseq60930',\n",
       " 'refseq80044',\n",
       " 'refseq35870',\n",
       " 'refseq42652',\n",
       " 'refseq20480',\n",
       " 'refseq19669',\n",
       " 'refseq159778',\n",
       " 'refseq167632',\n",
       " 'refseq123602',\n",
       " 'refseq12163',\n",
       " 'refseq120921',\n",
       " 'refseq81495',\n",
       " 'refseq83996',\n",
       " 'refseq22944',\n",
       " 'refseq55752',\n",
       " 'refseq15142',\n",
       " 'refseq21673',\n",
       " 'refseq17095',\n",
       " 'refseq54074',\n",
       " 'refseq200321',\n",
       " 'refseq79499',\n",
       " 'refseq30040',\n",
       " 'refseq182310',\n",
       " 'refseq106844',\n",
       " 'refseq100988',\n",
       " 'refseq47848',\n",
       " 'refseq38564',\n",
       " 'refseq42991',\n",
       " 'refseq183432',\n",
       " 'refseq52574',\n",
       " 'refseq1661',\n",
       " 'refseq199975',\n",
       " 'refseq182679',\n",
       " 'refseq167071',\n",
       " 'refseq18600',\n",
       " 'refseq121459',\n",
       " 'refseq179849',\n",
       " 'refseq35197',\n",
       " 'refseq62541',\n",
       " 'refseq63038',\n",
       " 'refseq49148',\n",
       " 'refseq14008',\n",
       " 'refseq111888',\n",
       " 'refseq164368',\n",
       " 'refseq122896',\n",
       " 'refseq11781',\n",
       " 'refseq115509',\n",
       " 'refseq49292',\n",
       " 'refseq166847',\n",
       " 'refseq15724',\n",
       " 'refseq719',\n",
       " 'refseq42364',\n",
       " 'refseq67973',\n",
       " 'refseq13764',\n",
       " 'refseq21428',\n",
       " 'refseq52578',\n",
       " 'refseq37115',\n",
       " 'refseq31434',\n",
       " 'refseq106806',\n",
       " 'refseq42248',\n",
       " 'refseq98356',\n",
       " 'refseq26362',\n",
       " 'refseq173762',\n",
       " 'refseq125950',\n",
       " 'refseq167631',\n",
       " 'refseq63080',\n",
       " 'refseq70437',\n",
       " 'refseq179803',\n",
       " 'refseq25962',\n",
       " 'refseq31135',\n",
       " 'refseq55020',\n",
       " 'refseq22153',\n",
       " 'refseq16380',\n",
       " 'refseq116296',\n",
       " 'refseq115570',\n",
       " 'refseq152086',\n",
       " 'refseq1756',\n",
       " 'refseq5677',\n",
       " 'refseq63195',\n",
       " 'refseq46988',\n",
       " 'refseq126088',\n",
       " 'refseq24931',\n",
       " 'refseq65960',\n",
       " 'refseq107755',\n",
       " 'refseq35660',\n",
       " 'refseq179934',\n",
       " 'refseq168882',\n",
       " 'refseq6480',\n",
       " 'refseq53986',\n",
       " 'refseq116325',\n",
       " 'refseq179646',\n",
       " 'refseq56919',\n",
       " 'refseq194616',\n",
       " 'refseq29950',\n",
       " 'refseq197795',\n",
       " 'refseq152991',\n",
       " 'refseq46445',\n",
       " 'refseq39765',\n",
       " 'refseq569',\n",
       " 'refseq190303',\n",
       " 'refseq42078',\n",
       " 'refseq47137',\n",
       " 'refseq27392',\n",
       " 'refseq45346',\n",
       " 'refseq116379',\n",
       " 'refseq177746',\n",
       " 'refseq43694',\n",
       " 'refseq88159',\n",
       " 'refseq127832',\n",
       " 'refseq56131',\n",
       " 'refseq25429',\n",
       " 'refseq67997',\n",
       " 'refseq118829',\n",
       " 'refseq116695',\n",
       " 'refseq196516',\n",
       " 'refseq56216',\n",
       " 'refseq35914',\n",
       " 'refseq54332',\n",
       " 'refseq28940',\n",
       " 'refseq116077',\n",
       " 'refseq138295',\n",
       " 'refseq54705',\n",
       " 'refseq122807',\n",
       " 'refseq42247',\n",
       " 'refseq24473',\n",
       " 'refseq27604',\n",
       " 'refseq42202',\n",
       " 'refseq98297',\n",
       " 'refseq37596',\n",
       " 'refseq60340',\n",
       " 'refseq86023',\n",
       " 'refseq81040',\n",
       " 'refseq31679',\n",
       " 'refseq48668',\n",
       " 'refseq32082',\n",
       " 'refseq173630',\n",
       " 'refseq165256',\n",
       " 'refseq165877',\n",
       " 'refseq35086',\n",
       " 'refseq82011',\n",
       " 'refseq111130',\n",
       " 'refseq78150',\n",
       " 'refseq53192',\n",
       " 'refseq177971',\n",
       " 'refseq116076',\n",
       " 'refseq50739',\n",
       " 'refseq40756',\n",
       " 'refseq32530',\n",
       " 'refseq110070',\n",
       " 'refseq42096',\n",
       " 'refseq166465',\n",
       " 'refseq124004',\n",
       " 'refseq104861',\n",
       " 'refseq185380',\n",
       " 'refseq199945',\n",
       " 'refseq186277',\n",
       " 'refseq35431',\n",
       " 'refseq51733',\n",
       " 'refseq22365',\n",
       " 'refseq166722',\n",
       " 'refseq23310',\n",
       " 'refseq21565',\n",
       " 'refseq167889',\n",
       " 'refseq198736',\n",
       " 'refseq180120',\n",
       " 'refseq10001',\n",
       " 'refseq54465',\n",
       " 'refseq38503',\n",
       " 'refseq53697',\n",
       " 'refseq95596',\n",
       " 'refseq124401',\n",
       " 'refseq116297',\n",
       " 'refseq62539',\n",
       " 'refseq16590',\n",
       " 'refseq187324',\n",
       " 'refseq26409',\n",
       " 'refseq10082',\n",
       " 'refseq143320',\n",
       " 'refseq43483',\n",
       " 'refseq119014',\n",
       " 'refseq16776',\n",
       " 'refseq15274',\n",
       " 'refseq116743',\n",
       " 'refseq13707',\n",
       " 'refseq17649',\n",
       " 'refseq120031',\n",
       " 'refseq173407',\n",
       " 'refseq59055',\n",
       " 'refseq4419',\n",
       " 'refseq62210',\n",
       " 'refseq165790',\n",
       " 'refseq58442',\n",
       " 'refseq199004',\n",
       " 'refseq47849',\n",
       " 'refseq35078',\n",
       " 'refseq203637',\n",
       " 'refseq55987',\n",
       " 'refseq48011',\n",
       " 'refseq31102',\n",
       " 'refseq106847',\n",
       " 'refseq36579',\n",
       " 'refseq164165',\n",
       " 'refseq15998',\n",
       " 'refseq85212',\n",
       " 'refseq606',\n",
       " 'refseq179885',\n",
       " 'refseq182384',\n",
       " 'refseq10231',\n",
       " 'refseq12642',\n",
       " 'refseq3737',\n",
       " 'refseq12157',\n",
       " 'refseq164598',\n",
       " 'refseq56547',\n",
       " 'refseq28716',\n",
       " 'refseq165782',\n",
       " 'refseq37525',\n",
       " 'refseq20256',\n",
       " 'refseq116388',\n",
       " 'refseq41473',\n",
       " 'refseq29953',\n",
       " 'refseq125921',\n",
       " 'refseq106146',\n",
       " 'refseq60969',\n",
       " 'refseq165902',\n",
       " 'refseq61252',\n",
       " 'refseq34030',\n",
       " 'refseq178915',\n",
       " 'refseq19867',\n",
       " 'refseq152902',\n",
       " 'refseq39445',\n",
       " 'refseq3584',\n",
       " 'refseq39278',\n",
       " 'refseq77266',\n",
       " 'refseq128846',\n",
       " 'refseq17419',\n",
       " 'refseq3972',\n",
       " 'refseq19900',\n",
       " 'refseq50029',\n",
       " 'refseq27574',\n",
       " 'refseq56165',\n",
       " 'refseq29159',\n",
       " 'refseq127834',\n",
       " 'refseq22039',\n",
       " 'refseq30666',\n",
       " 'refseq174496',\n",
       " 'refseq121789',\n",
       " 'refseq29593',\n",
       " 'refseq159623',\n",
       " 'refseq26463',\n",
       " 'refseq78118',\n",
       " 'refseq55892',\n",
       " 'refseq87410',\n",
       " 'refseq165792',\n",
       " 'refseq165970',\n",
       " 'refseq115938',\n",
       " 'refseq184376',\n",
       " 'refseq56640',\n",
       " 'refseq95602',\n",
       " 'refseq116728',\n",
       " 'refseq167581',\n",
       " 'refseq170309',\n",
       " 'refseq195544',\n",
       " 'refseq173628',\n",
       " 'refseq21809',\n",
       " 'refseq19457',\n",
       " 'refseq116316',\n",
       " 'refseq87462',\n",
       " 'refseq166721',\n",
       " 'refseq6698',\n",
       " 'refseq107841',\n",
       " 'refseq106208',\n",
       " 'refseq59245',\n",
       " 'refseq199908',\n",
       " 'refseq66624',\n",
       " 'refseq165897',\n",
       " 'refseq37990',\n",
       " 'refseq106673',\n",
       " 'refseq151305',\n",
       " 'refseq62214',\n",
       " 'refseq65043',\n",
       " 'refseq20517',\n",
       " 'refseq47859',\n",
       " 'refseq22460',\n",
       " 'refseq30207',\n",
       " 'refseq164242',\n",
       " 'refseq32886',\n",
       " 'refseq200322',\n",
       " 'refseq56797',\n",
       " 'refseq174384',\n",
       " 'refseq48595',\n",
       " 'refseq15539',\n",
       " 'refseq30893',\n",
       " 'refseq47137',\n",
       " 'refseq106548',\n",
       " 'refseq42602',\n",
       " 'refseq182458',\n",
       " 'refseq111884',\n",
       " 'refseq103338',\n",
       " 'refseq745',\n",
       " 'refseq115664',\n",
       " 'refseq163235',\n",
       " 'refseq49300',\n",
       " 'refseq129884',\n",
       " 'refseq97551',\n",
       " 'refseq146589',\n",
       " 'refseq29124',\n",
       " 'refseq128463',\n",
       " 'refseq26360',\n",
       " 'refseq132066',\n",
       " 'refseq168166',\n",
       " 'refseq106379',\n",
       " 'refseq163689',\n",
       " 'refseq58830',\n",
       " 'refseq128802',\n",
       " 'refseq4723',\n",
       " 'refseq185645',\n",
       " 'refseq33553',\n",
       " 'refseq89695',\n",
       " 'refseq184887',\n",
       " 'refseq32352',\n",
       " 'refseq117238',\n",
       " 'refseq116119',\n",
       " 'refseq128661',\n",
       " 'refseq116945',\n",
       " 'refseq53408',\n",
       " 'refseq56177',\n",
       " 'refseq82971',\n",
       " 'refseq128786',\n",
       " 'refseq74602',\n",
       " 'refseq142434',\n",
       " 'refseq129928',\n",
       " 'refseq87641',\n",
       " 'refseq180385',\n",
       " 'refseq18489',\n",
       " 'refseq23264',\n",
       " 'refseq7787',\n",
       " 'refseq35408',\n",
       " 'refseq7821',\n",
       " 'refseq180368',\n",
       " 'refseq24116',\n",
       " 'refseq123586',\n",
       " 'refseq198711',\n",
       " 'refseq198712',\n",
       " 'refseq23252',\n",
       " 'refseq22764',\n",
       " 'refseq180383',\n",
       " 'refseq107517',\n",
       " 'refseq135617',\n",
       " 'refseq20120',\n",
       " 'refseq47922',\n",
       " 'refseq181828',\n",
       " 'refseq11977',\n",
       " 'refseq33006',\n",
       " 'refseq46832',\n",
       " 'refseq9924',\n",
       " 'refseq180418',\n",
       " 'refseq130033',\n",
       " 'refseq18658',\n",
       " 'refseq180407',\n",
       " 'refseq181827',\n",
       " 'refseq78064',\n",
       " 'refseq4230',\n",
       " 'refseq34445',\n",
       " 'refseq58761',\n",
       " 'refseq24436',\n",
       " 'refseq158796',\n",
       " 'refseq40482',\n",
       " 'refseq37311',\n",
       " 'refseq117186',\n",
       " 'refseq187302',\n",
       " 'refseq93260',\n",
       " 'refseq50494',\n",
       " 'refseq172552',\n",
       " 'refseq180429',\n",
       " 'refseq203140',\n",
       " 'refseq180575',\n",
       " 'refseq4831',\n",
       " 'refseq11113',\n",
       " 'refseq181881',\n",
       " 'refseq129406',\n",
       " 'refseq130013',\n",
       " 'refseq41689',\n",
       " 'refseq8853',\n",
       " 'refseq46948',\n",
       " 'refseq85107',\n",
       " 'refseq130006',\n",
       " 'refseq49964',\n",
       " 'refseq21134',\n",
       " 'refseq34967',\n",
       " 'refseq130025',\n",
       " 'refseq14880',\n",
       " 'refseq181798',\n",
       " 'refseq24752',\n",
       " 'refseq6588',\n",
       " 'refseq181822',\n",
       " 'refseq130022',\n",
       " 'refseq4407',\n",
       " 'refseq130060',\n",
       " 'refseq39515',\n",
       " 'refseq52372',\n",
       " 'refseq114860',\n",
       " 'refseq22227',\n",
       " 'refseq181791',\n",
       " 'refseq181803',\n",
       " 'refseq86797',\n",
       " 'refseq181801',\n",
       " 'refseq123492',\n",
       " 'refseq30798',\n",
       " 'refseq5651',\n",
       " 'refseq2933',\n",
       " 'refseq89833',\n",
       " 'refseq181880',\n",
       " 'refseq22680',\n",
       " 'refseq4012',\n",
       " 'refseq45120',\n",
       " 'refseq189997',\n",
       " 'refseq53835',\n",
       " 'refseq130049',\n",
       " 'refseq184003',\n",
       " 'refseq118346',\n",
       " 'refseq78731',\n",
       " 'refseq116974',\n",
       " 'refseq29149',\n",
       " 'refseq29717',\n",
       " 'refseq25486',\n",
       " 'refseq64472',\n",
       " 'refseq178927',\n",
       " 'refseq2494',\n",
       " 'refseq16707',\n",
       " 'refseq32665',\n",
       " 'refseq9141',\n",
       " 'refseq117044',\n",
       " 'refseq128770',\n",
       " 'refseq51993',\n",
       " 'refseq49390',\n",
       " 'refseq14313',\n",
       " 'refseq12563',\n",
       " 'refseq116112',\n",
       " 'refseq151277',\n",
       " 'refseq169137',\n",
       " 'refseq26912',\n",
       " 'refseq165385',\n",
       " 'refseq4973',\n",
       " 'refseq110797',\n",
       " 'refseq10065',\n",
       " 'refseq15965',\n",
       " 'refseq118955',\n",
       " 'refseq21401',\n",
       " 'refseq186991',\n",
       " 'refseq16130',\n",
       " 'refseq13902',\n",
       " 'refseq6221',\n",
       " 'refseq194689',\n",
       " 'refseq182711',\n",
       " 'refseq94851',\n",
       " 'refseq96471',\n",
       " 'refseq147640',\n",
       " 'refseq166455',\n",
       " 'refseq165061',\n",
       " 'refseq78476',\n",
       " 'refseq48314',\n",
       " 'refseq18978',\n",
       " 'refseq9723',\n",
       " 'refseq55541',\n",
       " 'refseq126157',\n",
       " 'refseq125763',\n",
       " 'refseq148572',\n",
       " 'refseq57013',\n",
       " 'refseq55077',\n",
       " 'refseq144074',\n",
       " 'refseq12837',\n",
       " 'refseq29740',\n",
       " 'refseq111910',\n",
       " 'refseq89815',\n",
       " 'refseq33989',\n",
       " 'refseq68621',\n",
       " 'refseq62382',\n",
       " 'refseq164225',\n",
       " 'refseq166805',\n",
       " 'refseq36888',\n",
       " 'refseq28614',\n",
       " 'refseq168855',\n",
       " 'refseq60621',\n",
       " 'refseq30155',\n",
       " 'refseq24721',\n",
       " 'refseq125776',\n",
       " 'refseq15678',\n",
       " 'refseq60704',\n",
       " 'refseq104459',\n",
       " 'refseq178990',\n",
       " 'refseq68225',\n",
       " 'refseq145727',\n",
       " 'refseq166879',\n",
       " 'refseq49810',\n",
       " 'refseq7715',\n",
       " 'refseq106905',\n",
       " 'refseq26547',\n",
       " 'refseq131204',\n",
       " 'refseq152083',\n",
       " 'refseq107256',\n",
       " 'refseq49286',\n",
       " 'refseq164743',\n",
       " 'refseq5180',\n",
       " 'refseq4056',\n",
       " 'refseq54947',\n",
       " 'refseq66645',\n",
       " 'refseq65559',\n",
       " 'refseq128206',\n",
       " 'refseq166544',\n",
       " 'refseq70340',\n",
       " 'refseq163',\n",
       " 'refseq35634',\n",
       " 'refseq126842',\n",
       " 'refseq145103',\n",
       " 'refseq56709',\n",
       " 'refseq108190',\n",
       " 'refseq18560',\n",
       " 'refseq64266',\n",
       " 'refseq27164',\n",
       " 'refseq116206',\n",
       " 'refseq4551',\n",
       " 'refseq20825',\n",
       " 'refseq63881',\n",
       " 'refseq127785',\n",
       " 'refseq118950',\n",
       " 'refseq70131',\n",
       " 'refseq126144',\n",
       " 'refseq3110',\n",
       " 'refseq165791',\n",
       " 'refseq202280',\n",
       " 'refseq83466',\n",
       " 'refseq167578',\n",
       " 'refseq166624',\n",
       " 'refseq55960',\n",
       " 'refseq5374',\n",
       " 'refseq53645',\n",
       " 'refseq19112',\n",
       " 'refseq42064',\n",
       " 'refseq116773',\n",
       " 'refseq56946',\n",
       " 'refseq19899',\n",
       " 'refseq167537',\n",
       " 'refseq177981',\n",
       " 'refseq10159',\n",
       " 'refseq133550',\n",
       " 'refseq178908',\n",
       " 'refseq150818',\n",
       " 'refseq107022',\n",
       " 'refseq53471',\n",
       " 'refseq201610',\n",
       " 'refseq57599',\n",
       " 'refseq21795',\n",
       " 'refseq89823',\n",
       " 'refseq95',\n",
       " 'refseq728',\n",
       " 'refseq179027',\n",
       " 'refseq89260',\n",
       " 'refseq7187',\n",
       " 'refseq179810',\n",
       " 'refseq199967',\n",
       " 'refseq3754',\n",
       " 'refseq64835',\n",
       " 'refseq53764',\n",
       " 'refseq18140',\n",
       " 'refseq89749',\n",
       " 'refseq24953',\n",
       " 'refseq137781',\n",
       " 'refseq47125',\n",
       " 'refseq151131',\n",
       " 'refseq186246',\n",
       " 'refseq182875',\n",
       " 'refseq28514',\n",
       " 'refseq106410',\n",
       " 'refseq97089',\n",
       " 'refseq58294',\n",
       " 'refseq119340',\n",
       " 'refseq50016',\n",
       " 'refseq99475',\n",
       " 'refseq117731',\n",
       " 'refseq166739',\n",
       " 'refseq19965',\n",
       " 'refseq47396',\n",
       " 'refseq45598',\n",
       " 'refseq3094',\n",
       " 'refseq167082',\n",
       " 'refseq26387',\n",
       " 'refseq116594',\n",
       " 'refseq50577',\n",
       " 'refseq60712',\n",
       " 'refseq167039',\n",
       " 'refseq181594',\n",
       " 'refseq81272',\n",
       " 'refseq89486',\n",
       " 'refseq3386',\n",
       " 'refseq30',\n",
       " 'refseq128150',\n",
       " 'refseq1258',\n",
       " 'refseq1907',\n",
       " 'refseq21682',\n",
       " 'refseq6201',\n",
       " 'refseq54081',\n",
       " 'refseq19269',\n",
       " 'refseq3984',\n",
       " 'refseq14882',\n",
       " 'refseq96685',\n",
       " 'refseq36645',\n",
       " 'refseq15188',\n",
       " 'refseq53163',\n",
       " 'refseq27025',\n",
       " 'refseq62570',\n",
       " 'refseq96648',\n",
       " 'refseq11225',\n",
       " 'refseq20043',\n",
       " 'refseq145937',\n",
       " 'refseq118082',\n",
       " 'refseq6869',\n",
       " 'refseq56719',\n",
       " 'refseq50396',\n",
       " 'refseq62373',\n",
       " 'refseq70352',\n",
       " 'refseq104390',\n",
       " 'refseq53963',\n",
       " 'refseq6890',\n",
       " 'refseq24896',\n",
       " 'refseq23612',\n",
       " 'refseq14422',\n",
       " 'refseq99615',\n",
       " 'refseq7013',\n",
       " 'refseq169036',\n",
       " 'refseq167547',\n",
       " ...]"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_seq = most_frequent_16S_result.to_list()\n",
    "target_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "1eeac852",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2533"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "from pathlib import Path\n",
    "\n",
    "# 读取fasta文件\n",
    "def read_fasta(fasta_path):\n",
    "    records = list(SeqIO.parse(fasta_path, 'fasta'))\n",
    "    return records\n",
    "\n",
    "# 找出代表性序列用于建树\n",
    "def get_target_records(records):\n",
    "    keep_records = []\n",
    "    for record in records:\n",
    "        if record.id in target_seq:\n",
    "            keep_records.append(record)\n",
    "    return keep_records\n",
    "\n",
    "records = read_fasta(\"data/filtered_intergrated_16SrRNA.fasta\")\n",
    "target_records = get_target_records(records)\n",
    "\n",
    "# 保存序列\n",
    "output_file = Path('data/Fig1C_genus_rep_16SrRNA.fasta')\n",
    "SeqIO.write(target_records, output_file, 'fasta')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "6dff6b85",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_merged_df.to_csv(\"data/TableS1_HQ_filtered16S_lineage.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "psmpa-ai",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
